content/docs/projects/tdd.md

```diff
@@ -0,0 +1,602 @@
+3. <span style="color: blue;">Refactor</span>: Once the code passes the test, you can refactor the code to improve its design, readability, and performance. This is done without changing the functionality of the code, so the tests should still pass.
+
+The red-green-refactor approach is useful because it helps you focus on one small piece of functionality at a time, which makes the development process more manageable. It also encourages you to think about the design of your code upfront, and helps ensure that your code is well-tested.
+
+#### 5.2.1.2 Example
+
+Here is an example of the red, green, refactor approach in test-driven development:
+
+1. <span style="color: red;">Red</span>: Write a test to check that a function called `add` can be used to add two numbers together and return the result. The test should define the expected behavior of the `add` function, and should be written before the `add` function itself is written.
+
+```python
+def test_add():
+  result = add(2, 3)
+  assert result == 5
+```
+
+2. <span style="color: green;">Green</span>: Write the `add` function to make the test pass. This should be done in a way that satisfies the requirements of the test and ensures that the `add` function works as expected.
+
+```python
+def add(a, b):
+  return 5
+```
+
+3. <span style="color: blue;">Refactor</span>: Once the `add` function passes the test, you can refactor the code to make it more modular and readable. This might involve restructuring the code, adding comments, or making other improvements.
+
+```python
+def add(a, b):
+  # Return the sum of a and b
+  return a + b
+```
+
+In this example, the red, green, refactor approach is used to write a simple function called `add` that can be used to add two numbers together and return the result. The test is written first, and then the code is written to make the test pass. Finally, the code is refactored to improve its readability and maintainability.
+
+### 5.2.2 Basic Rules
+
+There are several basic rules that you should follow when writing unit tests in Frappe:
+
+1. Each test case should be a subclass of `unittest.TestCase` and should contain one or more test methods that begin with the word `test`.
+2. Test methods should use the various assertions provided by the `unittest` module to verify that the code under test is working correctly.
+3. Test cases should be isolated from one another. This means that each test should run independently and not depend on any other test to be run first.
+4. Test cases should be repeatable. This means that the same test should produce the same results every time it is run, regardless of the environment or other factors.
+5. Test cases should be thorough. This means that they should cover all relevant scenarios and test for all possible outcomes.
+6. Test cases should be easy to read and understand. This means that they should be well-structured and use clear, descriptive names for variables and methods.
```

Comment:
Additional reminder: test code is still code. So it should follow the same standards as production code (i.e following DRY and KISS principles).

Reply:
856cf12959eea44b2307f78bbf6ce9b83a24a51b